src/dune_rules/modules.ml

```diff
@@ -211,177 +221,291 @@ module Mangle = struct
       let path = Module.path m in
       let prefix = prefix t in
       match t with
-      | Exe | Melange -> prefix.public :: path
+      | Exe | Melange -> (Option.value_exn prefix).public :: path
+      | Unwrapped ->
+        if is_lib_interface then List.rev path |> List.tl |> List.rev else path
       | Lib _ ->
         let path =
           if is_lib_interface then List.rev path |> List.tl |> List.rev
           else path
         in
-        Visibility.Map.find prefix (Module.visibility m) :: path
+        Visibility.Map.find (Option.value_exn prefix) (Module.visibility m)
+        :: path
     in
     Module.set_obj_name m (Module_name.Path.wrap path_for_mangle)
 end
 
-let impl_only_of_map m =
-  Module_name.Map.fold m ~init:[] ~f:(fun m acc ->
-      if Module.has m ~ml_kind:Impl then m :: acc else acc)
-
-module Wrapped = struct
-  module Group = struct
-    type t =
-      { alias : Module.t
-      ; modules : node Module_name.Map.t
-      ; name : Module_name.t
-      }
+module Group = struct
+  type t =
+    { alias : Module.t option
+    ; modules : node Module_name.Map.t
+    ; name : Module_name.t option
+          (* actually, only the top most [t] might have name set to [None]. But
+              writing this in a type safe way would add a lot of boilerplate *)
+    }
```

Comment:
this type would be precise:

```
type subgroup =
    { alias : Module.t
     ; modules : node Module_name.Map.t
     ; name : Module_name.t
     }
and 'a node = Group of t | Module of Module.t
type wrapped = subgroup
type unwrapped = node Module_name.Map.t
```

although it doesn't seem to be worth the hassle as it will require me to write a lot of functions twice.